activesupport/lib/active_support/notifications/fanout.rb

```diff
@@ -24,12 +24,15 @@ def initialize
       def subscribe(pattern = nil, callable = nil, monotonic: false, &block)
         subscriber = Subscribers.new(pattern, callable || block, monotonic)
         synchronize do
-          if String === pattern
+          case pattern
+          when String
             @string_subscribers[pattern] << subscriber
             @listeners_for.delete(pattern)
-          else
+          when NilClass, Regexp
             @other_subscribers << subscriber
             @listeners_for.clear
+          else
+            raise ArgumentError,  "pattern must be specified as a String, Regexp or empty"
```

Comment:
I think this check should happen much earlier so that we don't have to create a `Subscribers` object when the argument isn't valid.

Reply:
This will not be a hot-path and will immediately fail, so doesn't seem like ideal to move params check before others as an extra step.

Reply:
That is true, thank you for clarifying :+1: